File: frontend/src/models/data/Tree.ts
import DataDivider from "./DataDivider";
import EndTreeNode from "./EndTreeNode";
import MiddleTreeNode from "./MiddleTreeNode";
import TreeNode from "./TreeNode"

export default class Tree {
    root: TreeNode;
    endNodeCount: number;
    constructor() {
        this.endNodeCount = 2;
        this.root = new MiddleTreeNode("0");
        (this.root as MiddleTreeNode).setChild(new EndTreeNode("A","00"));
        (this.root as MiddleTreeNode).setChild(new EndTreeNode("B", "01"));
    };

    getNodeByKey(key: string) {
        if (key == "0") return this.root;

        let node = this.root;
        for (let i = 1; i < key.length; i++) {
            const nextKey = key.substring(0, i + 1);
            //if (node instanceof EndTreeNode) return null;
            if (!(node as MiddleTreeNode).next.has(nextKey)) return null;
            node = (node as MiddleTreeNode).next.get(nextKey)!; // ! - non-null assertion operator
        }
        return node;
    };

    addNode(key: string) {
   
        if (this.endNodeCount == 5) return;
        const node = this.getNodeByKey(key);
        if (node == null) return;
        if (node instanceof EndTreeNode) {
            const aboveNode = this.getNodeByKey(key.substring(0, key.length-1))!;
            const newNode = new MiddleTreeNode(key);
            newNode.setChild(new EndTreeNode((node as EndTreeNode).value, key + "0"));
            newNode.setChild(new EndTreeNode("ABCDE"[this.endNodeCount], key + "1"));
            (aboveNode as MiddleTreeNode).setChild(newNode);
            this.endNodeCount++;
        }

    };

    removeNode(key: string) {
        if (this.endNodeCount == 2) return;
        const node = this.getNodeByKey(key);
        if (node == null) return;
        if (node instanceof MiddleTreeNode) return;

        const aboveNode = this.getNodeByKey(key.substring(0, key.length - 1))!;
        (aboveNode as MiddleTreeNode).removeChild(this.getNodeByKey(key)!);
        const aboveAboveNode = this.getNodeByKey(key.substring(0, key.length - 2))!;
        (aboveAboveNode as MiddleTreeNode).setChild(new EndTreeNode((this.getNodeByKey(aboveNode + "0") as EndTreeNode)!.value, aboveNode.key))
        this.endNodeCount--;
    };

    changeDivider(key: string, divider: DataDivider) {
        (this.getNodeByKey(key) as MiddleTreeNode).divider = divider;
    }

}

File: backend/Models/Tree.cs
using Newtonsoft.Json.Linq;

namespace Graphic_Neural_Network.backend.Models
{
    public class Tree
    {
        public TreeNode Root { get; set; }
        public int EndNodeCount { get; set; }

        public Tree()
        {
            EndNodeCount = 2;
            Root = new MiddleTreeNode("0");
            ((MiddleTreeNode)Root).SetChild(new EndTreeNode("A", "00"));
            ((MiddleTreeNode)Root).SetChild(new EndTreeNode("B", "01"));
        }

        public virtual JObject ToJson()
        {
            var jo = new JObject();
            jo["EndNodeCount"] = this.EndNodeCount;
            jo["root"] = this.Root.ToJson();
            return jo;
        }

        public TreeNode GetNodeByKey(string key)
        {
            if (key == "0") return Root;

            TreeNode node = Root;

            for (int i = 1; i < key.Length; i++)
            {
                string nextKey = key.Substring(0, i + 1);
                if (node is not MiddleTreeNode middleNode || !middleNode.Next.ContainsKey(nextKey))
                    return null;

                node = middleNode.Next[nextKey];
            }

            return node;
        }

        public void AddNode(string key)
        {
            if (EndNodeCount == 5) return;

            TreeNode node = GetNodeByKey(key);
            if (node == null || node is not EndTreeNode endNode) return;

            TreeNode aboveNode = GetNodeByKey(key[..^1]);
            if (aboveNode is not MiddleTreeNode aboveMiddle) return;

            MiddleTreeNode newNode = new(key);
            newNode.SetChild(new EndTreeNode(endNode.Value, key + "0"));
            newNode.SetChild(new EndTreeNode("ABCDE"[EndNodeCount].ToString(), key + "1"));
            aboveMiddle.SetChild(newNode);

            EndNodeCount++;
        }

        public void RemoveNode(string key)
        {
            if (EndNodeCount == 2) return;

            TreeNode node = GetNodeByKey(key);
            if (node == null || node is MiddleTreeNode) return;

            string parentKey = key[..^1];
            TreeNode parentNode = GetNodeByKey(parentKey);
            if (parentNode is not MiddleTreeNode parentMiddle) return;

            parentMiddle.RemoveChild(node);

            string grandParentKey = key[..^2];
            TreeNode grandParent = GetNodeByKey(grandParentKey);
            if (grandParent is not MiddleTreeNode grandMiddle) return;

            string siblingKey = parentKey + "0";
            EndTreeNode sibling = GetNodeByKey(siblingKey) as EndTreeNode;
            if (sibling == null) return;

            grandMiddle.SetChild(new EndTreeNode(sibling.Value, parentKey));
            EndNodeCount--;
        }

        public void ChangeDivider(string key, DataDivider divider)
        {
            if (GetNodeByKey(key) is MiddleTreeNode middle)
                middle.SetDivider(divider);
        }

        public DataDivider GetDivider(string key)
        {
            return GetNodeByKey(key) is MiddleTreeNode middle ? middle.Divider : null;
        }
    }

}
